File: app/src/main/java/com/drdisagree/iconify/ui/fragments/Changelog.java
package com.drdisagree.iconify.ui.fragments;

import static com.drdisagree.iconify.common.Const.CHANGELOG_URL;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.text.style.URLSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.text.HtmlCompat;

import com.drdisagree.iconify.BuildConfig;
import com.drdisagree.iconify.Iconify;
import com.drdisagree.iconify.R;
import com.drdisagree.iconify.databinding.FragmentChangelogBinding;
import com.drdisagree.iconify.ui.base.BaseFragment;
import com.drdisagree.iconify.ui.dialogs.LoadingDialog;
import com.drdisagree.iconify.ui.utils.ViewHelper;
import com.drdisagree.iconify.utils.extension.TaskExecutor;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Changelog extends BaseFragment {

    private FragmentChangelogBinding binding;
    private GrabChangelog grabChangelog = null;

    public static String usernameToLink(String str) {
        String regexPattern = "@([A-Za-z\\d_-]+)";
        Pattern pattern = Pattern.compile(regexPattern);
        Matcher matcher = pattern.matcher(str);

        StringBuffer sb = new StringBuffer();
        while (matcher.find()) {
            String username = matcher.group(1);
            String link = "<a href=\"https://github.com/" + username + "\">@" + username + "</a>";
            matcher.appendReplacement(sb, link);
        }
        matcher.appendTail(sb);

        return sb.toString();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentChangelogBinding.inflate(inflater, container, false);
        View view = binding.getRoot();

        // Header
        ViewHelper.setHeader(requireContext(), getParentFragmentManager(), binding.header.toolbar, R.string.activity_title_changelog);

        try {
            grabChangelog = new GrabChangelog();
            grabChangelog.execute();
        } catch (Exception ignored) {
        }

        return view;
    }

    @Override
    public void onPause() {
        if (grabChangelog != null) grabChangelog.cancel(true);
        super.onPause();
    }

    @Override
    public void onStop() {
        if (grabChangelog != null) grabChangelog.cancel(true);
        super.onStop();
    }

    @Override
    public void onDestroy() {
        if (grabChangelog != null) grabChangelog.cancel(true);
        super.onDestroy();
    }

    @SuppressLint("StaticFieldLeak")
    private class GrabChangelog extends TaskExecutor<Integer, Integer, JSONObject> {

        LoadingDialog loadingDialog = new LoadingDialog(requireContext());
        boolean connectionAvailable = false;

        @Override
        protected void onPreExecute() {
            loadingDialog.show(Iconify.Companion.getAppContextLocale().getResources().getString(R.string.loading_dialog_wait), true);
        }

        @Override
        protected JSONObject doInBackground(Integer... integers) {
            try {
                URL myUrl = new URL(CHANGELOG_URL + BuildConfig.VERSION_NAME);
                URLConnection connection = myUrl.openConnection();
                connection.setConnectTimeout(5000);
                connection.connect();
                connectionAvailable = true;
            } catch (Exception e) {
                connectionAvailable = false;
            }

            StringBuilder release_note = null;

            if (connectionAvailable) {
                String parseChangelog = CHANGELOG_URL + BuildConfig.VERSION_NAME;
                HttpURLConnection urlConnection = null;
                BufferedReader bufferedReader = null;

                try {
                    URL url = new URL(parseChangelog);
                    urlConnection = (HttpURLConnection) url.openConnection();
                    urlConnection.connect();

                    InputStream inputStream = urlConnection.getInputStream();

                    bufferedReader = new BufferedReader(new InputStreamReader(inputStream));

                    release_note = new StringBuilder();

                    String line;

                    while ((line = bufferedReader.readLine()) != null) {
                        release_note.append(line);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    release_note = null;
                } finally {
                    if (urlConnection != null) {
                        urlConnection.disconnect();
                    }
                    if (bufferedReader != null) {
                        try {
                            bufferedReader.close();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }

            if (release_note == null)
                release_note = new StringBuilder(getResources().getString(R.string.individual_changelog_not_found));

            try {
                return new JSONObject(release_note.toString());
            } catch (JSONException e) {
                return null;
            }
        }

        @Override
        protected void onPostExecute(JSONObject release_note) {
            loadingDialog.hide();
            String title, changes;

            if (connectionAvailable) {
                if (release_note == null || release_note.toString().length() == 0) {
                    title = getResources().getString(R.string.changelog_not_found);
                    changes = "";
                } else {
                    try {
                        String data = release_note.getString("body");
                        title = data.substring(0, data.indexOf("\r\n\r\n"));
                        changes = data.substring(data.indexOf("\n##")).substring(1);

                        title = title.replace("### ", "");
                        changes = usernameToLink(changes.replace("## ", "<b>").replace(":\r\n", ":</b><br>").replace("- __", "<br><b>• ").replace("__\r\n", "</b><br>").replace("    - ", "&emsp;◦ ").replace("- ", "• ").replace("\r\n", "<br>"));
                    } catch (JSONException e) {
                        title = getResources().getString(R.string.changelog_not_found);
                        changes = "";
                    }
                }
            } else {
                title = getResources().getString(R.string.no_internet_connection);
                changes = "";
            }

            binding.changelogTitle.setText(HtmlCompat.fromHtml(title, HtmlCompat.FROM_HTML_MODE_LEGACY));
            binding.changelogText.setText(HtmlCompat.fromHtml(changes, HtmlCompat.FROM_HTML_MODE_LEGACY));

            if (Objects.equals(changes, "")) binding.changelogText.setVisibility(View.GONE);
            else {
                SpannableString spannableString = new SpannableString(HtmlCompat.fromHtml(changes, HtmlCompat.FROM_HTML_MODE_LEGACY));
                URLSpan[] urls = spannableString.getSpans(0, spannableString.length(), URLSpan.class);

                for (URLSpan urlSpan : urls) {
                    ClickableSpan clickableSpan = new ClickableSpan() {
                        @Override
                        public void onClick(@NonNull View view) {
                            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(urlSpan.getURL()));
                            startActivity(intent);
                        }
                    };

                    int start = spannableString.getSpanStart(urlSpan);
                    int end = spannableString.getSpanEnd(urlSpan);
                    spannableString.setSpan(clickableSpan, start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                    spannableString.removeSpan(urlSpan);
                }

                binding.changelogText.setText(spannableString);
                binding.changelogText.setMovementMethod(LinkMovementMethod.getInstance());
            }

            binding.changelog.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/com/drdisagree/iconify/ui/fragments/settings/Changelog.kt
package com.drdisagree.iconify.ui.fragments.settings

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Spannable
import android.text.SpannableString
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.style.URLSpan
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.text.HtmlCompat
import com.drdisagree.iconify.BuildConfig
import com.drdisagree.iconify.Iconify.Companion.appContextLocale
import com.drdisagree.iconify.R
import com.drdisagree.iconify.common.Const.CHANGELOG_URL
import com.drdisagree.iconify.databinding.FragmentChangelogBinding
import com.drdisagree.iconify.ui.base.BaseFragment
import com.drdisagree.iconify.ui.dialogs.LoadingDialog
import com.drdisagree.iconify.ui.utils.ViewHelper.setHeader
import com.drdisagree.iconify.utils.extension.TaskExecutor
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.regex.Pattern

class Changelog : BaseFragment() {

    private lateinit var binding: FragmentChangelogBinding
    private var grabChangelog: GrabChangelog? = null
    private var tag = Changelog::class.java.simpleName

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentChangelogBinding.inflate(inflater, container, false)
        val view: View = binding.getRoot()

        // Header
        setHeader(
            requireContext(),
            getParentFragmentManager(),
            binding.header.toolbar,
            R.string.activity_title_changelog
        )

        try {
            grabChangelog = GrabChangelog()
            grabChangelog!!.execute()
        } catch (ignored: Exception) {
        }

        return view
    }

    private inner class GrabChangelog : TaskExecutor<Int?, Int?, JSONObject?>() {

        var loadingDialog = LoadingDialog(requireContext())
        var connectionAvailable = false

        override fun onPreExecute() {
            loadingDialog.show(
                appContextLocale.resources.getString(R.string.loading_dialog_wait),
                true
            )
        }

        override fun doInBackground(vararg params: Int?): JSONObject? {
            connectionAvailable = try {
                val myUrl = URL(CHANGELOG_URL + BuildConfig.VERSION_NAME)
                val connection = myUrl.openConnection()
                connection.setConnectTimeout(5000)
                connection.connect()

                true
            } catch (e: Exception) {
                false
            }

            var releaseNote: StringBuilder? = null

            if (connectionAvailable) {
                val parseChangelog: String = CHANGELOG_URL + BuildConfig.VERSION_NAME
                var urlConnection: HttpURLConnection? = null
                var bufferedReader: BufferedReader? = null

                try {
                    val url = URL(parseChangelog)
                    urlConnection = url.openConnection() as HttpURLConnection
                    urlConnection.connect()

                    val inputStream = urlConnection.inputStream
                    bufferedReader = BufferedReader(InputStreamReader(inputStream))
                    releaseNote = StringBuilder()

                    var line: String?
                    while (bufferedReader.readLine().also { line = it } != null) {
                        releaseNote.append(line)
                    }
                } catch (e: Exception) {
                    Log.e(tag, e.toString())
                    releaseNote = null
                } finally {
                    urlConnection?.disconnect()

                    if (bufferedReader != null) {
                        try {
                            bufferedReader.close()
                        } catch (e: IOException) {
                            Log.e(tag, e.toString())
                        }
                    }
                }
            }

            if (releaseNote == null) releaseNote =
                StringBuilder(resources.getString(R.string.individual_changelog_not_found))

            return try {
                JSONObject(releaseNote.toString())
            } catch (e: JSONException) {
                null
            }
        }

        override fun onPostExecute(result: JSONObject?) {
            loadingDialog.hide()

            var title: String
            var changes: String

            if (connectionAvailable) {
                if (result == null || result.toString().isEmpty()) {
                    title = resources.getString(R.string.changelog_not_found)
                    changes = ""
                } else {
                    try {
                        val data = result.getString("body")

                        title = data.substring(0, data.indexOf("\r\n\r\n"))
                        changes = data.substring(data.indexOf("\n##")).substring(1)

                        title = title.replace("### ", "")
                        changes = usernameToLink(
                            changes.replace("## ", "<b>")
                                .replace(":\r\n", ":</b><br>")
                                .replace("- __", "<br><b>• ")
                                .replace("__\r\n", "</b><br>")
                                .replace("    - ", "&emsp;◦ ")
                                .replace("- ", "• ")
                                .replace("\r\n", "<br>")
                        )
                    } catch (e: JSONException) {
                        title = resources.getString(R.string.changelog_not_found)
                        changes = ""
                    }
                }
            } else {
                title = resources.getString(R.string.no_internet_connection)
                changes = ""
            }

            binding.changelogTitle.text = HtmlCompat.fromHtml(
                title,
                HtmlCompat.FROM_HTML_MODE_LEGACY
            )
            binding.changelogText.text = HtmlCompat.fromHtml(
                changes,
                HtmlCompat.FROM_HTML_MODE_LEGACY
            )

            if (changes == "") binding.changelogText.visibility = View.GONE else {
                val spannableString = SpannableString(
                    HtmlCompat.fromHtml(
                        changes,
                        HtmlCompat.FROM_HTML_MODE_LEGACY
                    )
                )
                val urls = spannableString.getSpans(
                    0,
                    spannableString.length,
                    URLSpan::class.java
                )

                for (urlSpan in urls) {
                    val clickableSpan: ClickableSpan = object : ClickableSpan() {
                        override fun onClick(view: View) {
                            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(urlSpan.url))
                            startActivity(intent)
                        }
                    }

                    val start = spannableString.getSpanStart(urlSpan)
                    val end = spannableString.getSpanEnd(urlSpan)

                    spannableString.setSpan(
                        clickableSpan,
                        start,
                        end,
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )

                    spannableString.removeSpan(urlSpan)
                }

                binding.changelogText.text = spannableString
                binding.changelogText.movementMethod = LinkMovementMethod.getInstance()
            }

            binding.changelog.visibility = View.VISIBLE
        }
    }

    override fun onPause() {
        grabChangelog?.cancel(true)

        super.onPause()
    }

    override fun onStop() {
        grabChangelog?.cancel(true)

        super.onStop()
    }

    override fun onDestroy() {
        grabChangelog?.cancel(true)

        super.onDestroy()
    }

    companion object {
        fun usernameToLink(str: String): String {
            val regexPattern = "@([A-Za-z\\d_-]+)"
            val pattern = Pattern.compile(regexPattern)
            val matcher = pattern.matcher(str)
            val sb = StringBuffer()

            while (matcher.find()) {
                val username = matcher.group(1)
                val link = "<a href=\"https://github.com/$username\">@$username</a>"
                matcher.appendReplacement(sb, link)
            }

            matcher.appendTail(sb)

            return sb.toString()
        }
    }
}